Quake/pr_ext.c

```diff
@@ -354,6 +354,232 @@ static void PF_anglemod(void)
+static qboolean QC_FixFileName(const char* name, const char** result, const char** fallbackread)
+{
+	if (!*name ||	//blank names are bad
+		strchr(name, ':') ||	//dos/win absolute path, ntfs ADS, amiga drives. reject them all.
+		strchr(name, '\\') ||	//windows-only paths.
+		*name == '/' ||	//absolute path was given - reject
+		strstr(name, ".."))	//someone tried to be clever.
+	{
+		return false;
+	}
+
+	*fallbackread = name;
+	//if its a user config, ban any fallback locations so that csqc can't read passwords or whatever.
+	if ((!strchr(name, '/') || q_strncasecmp(name, "configs/", 8))
+		&& !q_strcasecmp(COM_FileGetExtension(name), "cfg")
+		&& q_strncasecmp(name, "particles/", 10) && q_strncasecmp(name, "huds/", 5) && q_strncasecmp(name, "models/", 7))
+		*fallbackread = NULL;
+	*result = va("data/%s", name);
+	return true;
+}
+
+//small note on access modes:
+//when reading, we fopen files inside paks, for compat with (crappy non-zip-compatible) filesystem code
+//when writing, we directly fopen the file such that it can never be inside a pak.
+//this means that we need to take care when reading in order to detect EOF properly.
+//writing doesn't need anything like that, so it can just dump stuff out, but we do need to ensure that the modes don't get mixed up, because trying to read from a writable file will not do what you would expect.
+//even libc mandates a seek between reading+writing, so no great loss there.
+static struct qcfile_s
+{
+	qcvm_t* owningvm;
+	char cache[1024];
+	int cacheoffset, cachesize;
+	FILE* file;
+	int fileoffset;
+	int filesize;
+	int filebase;	//the offset of the file inside a pak
+	int mode;
+} *qcfiles;
+static size_t qcfiles_max;
+#define QC_FILE_BASE 1
+static void PF_fopen(void)
+{
+	const char* fname = G_STRING(OFS_PARM0);
+	int fmode = G_FLOAT(OFS_PARM1);
+	const char* fallback;
+	FILE* file;
+	size_t i;
+	char name[MAX_OSPATH], * sl;
+	int filesize = 0;
+
+	G_FLOAT(OFS_RETURN) = -1;	//assume failure
+
+	if (!QC_FixFileName(fname, &fname, &fallback))
+	{
+		Con_Printf("qcfopen: Access denied: %s\n", fname);
+		return;
+	}
+	//if we were told to use 'foo.txt'
+	//fname is now 'data/foo.txt'
+	//fallback is now 'foo.txt', and must ONLY be read.
+
+	switch (fmode)
+	{
+	case 0: //read
+		filesize = COM_FOpenFile(fname, &file, NULL);
+		if (!file && fallback)
+			filesize = COM_FOpenFile(fallback, &file, NULL);
+		break;
+	case 1:	//append
+		q_snprintf(name, sizeof(name), "%s/%s", com_gamedir, fname);
+		Sys_mkdir(name);
+		file = fopen(name, "w+b");
+		if (file)
+			fseek(file, 0, SEEK_END);
+		break;
+	case 2: //write
+		q_snprintf(name, sizeof(name), "%s/%s", com_gamedir, fname);
+		sl = name + 1;
+		while (*sl)
+		{
+			if (*sl == '/')
+			{
+				*sl = 0;
+				Sys_mkdir(name);	//make sure each part of the path exists.
+				*sl = '/';
+			}
+			sl++;
+		}
+		file = fopen(name, "wb");
+		break;
+	default:
+		Con_Warning("PF_fopen: couldn't open file\n");
```

Comment:
Wrong / misleading warning for unsupported file open mode.

Reply:
Besides, the `file` variable will contain garbage and shall not be NULL if we hit this: either return here instead of break, or initialize `file` to NULL at top level.